### PR TITLE
filterx: `cached_json_file()`

### PR DESCRIPTION
**Pull request by alltilla**

### CODE REVIEW

File: modules/json/json-plugin.c [hunk 1]
```diff
@@ -42,6 +43,12 @@ static Plugin json_plugins[] =
     .name = "format_json",
     .construct = filterx_format_json_new_construct,
   },
+  {
+    .type = LL_CONTEXT_FILTERX_FUNC,
+    .name = "cached_json_file",
```
Comment (bazsi) [inline modules/json/json-plugin.c:48]: I'd probably rename this, if it's not a big hassle. We have started to write functions a'la parse_kv() etc, so it should be `cache_json_file()` or something like that.

Similarly, if we were to write a simple json parsing function, that would be parse_json()

File: modules/json/filterx-cached-json-file.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ */
+
+#include "filterx-cached-json-file.h"
+#include "filterx/object-json.h"
+
+typedef struct FilterXFunctionCachedJsonFile_
+{
+  FilterXFunction super;
+  FilterXObject *cached_json;
```
Comment (bazsi) [inline modules/json/filterx-cached-json-file.c:56]: This needs to be frozen, as it is accessed by multiple threads and the refcounts are not atomic.

File: lib/filterx/expr-getattr.c [hunk 3]
```diff
@@ -52,6 +52,30 @@ _eval(FilterXExpr *s)
...
+{
+  FilterXGetAttr *self = (FilterXGetAttr *) s;
+
+  gboolean result = FALSE;
+
+  FilterXObject *variable = filterx_expr_eval_typed(self->operand);
+  if (!variable)
+    return FALSE;
+
+  if (variable->readonly)
```
Comment (bazsi) [inline lib/filterx/expr-getattr.c:66]: we don't have readonly yet at this point in the history. That part should probably be bubbled up to the next patch.

File: modules/json/filterx-cached-json-file.c (hunk 2 above)
Comment (bazsi) [reply modules/json/filterx-cached-json-file.c:56]: I've fixed this with a force push to this branch.

File: lib/filterx/expr-getattr.c (hunk 3 above)
Comment (bazsi) [reply lib/filterx/expr-getattr.c:66]: I've fixed this up with a forced push to this branch.

**Comment by bazsi** [comment]
Probably not because of this, but there's a circular reference between the json_object and the cached filterx object when we query any value from within the cached JSON.

Reproduction:
```
@version: current

log {
	source { tcp(port(2000) flags(no-parse)); };

	filterx {
		x = cached_json_file("/install/proba.json");
		x.sudo;
	};
	destination { file("/dev/null"); };
};

```

JSON file:
```
{"sudo":{"accept":{"uuid":"b5fa4071d2-11a8-4ea9-244d-b47199c74c","server_time":{"seconds":1681742321,"nanoseconds":112963345,"iso8601":"20230417143841Z","localtime":"Apr 17 16:38:41"},"submit_time":{"seconds":1681742321,"nanoseconds":96494197,"iso8601":"20230417143841Z","localtime":"Apr 17 16:38:41"},"submituser":"root","command":"/usr/bin/echo","runuser":"root","runcwd":"/root","ttyname":"/dev/pts/2","submithost":"localhost.localdomain","submitcwd":"/root","runuid":0,"columns":118,"lines":60,"runargv":["echo","bla"],"runenv":["LS_COLORS=rs=0:di=38;5;33:ln=38;5;51:mh=00:pi=40;38;5;11:so=38;5;13:do=38;5;5:bd=48;5;232;38;5;11:cd=48;5;232;38;5;3:or=48;5;232;38;5;9:mi=01;05;37;41:su=48;5;196;38;5;15:sg=48;5;11;38;5;16:ca=48;5;196;38;5;226:tw=48;5;10;38;5;16:ow=48;5;10;38;5;21:st=48;5;21;38;5;15:ex=38;5;40:*.tar=38;5;9:*.tgz=38;5;9:*.arc=38;5;9:*.arj=38;5;9:*.taz=38;5;9:*.lha=38;5;9:*.lz4=38;5;9:*.lzh=38;5;9:*.lzma=38;5;9:*.tlz=38;5;9:*.txz=38;5;9:*.tzo=38;5;9:*.t7z=38;5;9:*.zip=38;5;9:*.z=38;5;9:*.dz=38;5;9:*.gz=38;5;9:*.lrz=38;5;9:*.lz=38;5;9:*.lzo=38;5;9:*.xz=38;5;9:*.zst=38;5;9:*.tzst=38;5;9:*.bz2=38;5;9:*.bz=38;5;9:*.tbz=38;5;9:*.tbz2=38;5;9:*.tz=38;5;9:*.deb=38;5;9:*.rpm=38;5;9:*.jar=38;5;9:*.war=38;5;9:*.ear=38;5;9:*.sar=38;5;9:*.rar=38;5;9:*.alz=38;5;9:*.ace=38;5;9:*.zoo=38;5;9:*.cpio=38;5;9:*.7z=38;5;9:*.rz=38;5;9:*.cab=38;5;9:*.wim=38;5;9:*.swm=38;5;9:*.dwm=38;5;9:*.esd=38;5;9:*.jpg=38;5;13:*.jpeg=38;5;13:*.mjpg=38;5;13:*.mjpeg=38;5;13:*.gif=38;5;13:*.bmp=38;5;13:*.pbm=38;5;13:*.pgm=38;5;13:*.ppm=38;5;13:*.tga=38;5;13:*.xbm=38;5;13:*.xpm=38;5;13:*.tif=38;5;13:*.tiff=38;5;13:*.png=38;5;13:*.svg=38;5;13:*.svgz=38;5;13:*.mng=38;5;13:*.pcx=38;5;13:*.mov=38;5;13:*.mpg=38;5;13:*.mpeg=38;5;13:*.m2v=38;5;13:*.mkv=38;5;13:*.webm=38;5;13:*.ogm=38;5;13:*.mp4=38;5;13:*.m4v=38;5;13:*.mp4v=38;5;13:*.vob=38;5;13:*.qt=38;5;13:*.nuv=38;5;13:*.wmv=38;5;13:*.asf=38;5;13:*.rm=38;5;13:*.rmvb=38;5;13:*.flc=38;5;13:*.avi=38;5;13:*.fli=38;5;13:*.flv=38;5;13:*.gl=38;5;13:*.dl=38;5;13:*.xcf=38;5;13:*.xwd=38;5;13:*.yuv=38;5;13:*.cgm=38;5;13:*.emf=38;5;13:*.ogv=38;5;13:*.ogx=38;5;13:*.aac=38;5;45:*.au=38;5;45:*.flac=38;5;45:*.m4a=38;5;45:*.mid=38;5;45:*.midi=38;5;45:*.mka=38;5;45:*.mp3=38;5;45:*.mpc=38;5;45:*.ogg=38;5;45:*.ra=38;5;45:*.wav=38;5;45:*.oga=38;5;45:*.opus=38;5;45:*.spx=38;5;45:*.xspf=38;5;45:","LANG=en_US.UTF-8","HOSTNAME=localhost.localdomain","TERM=xterm-256color","PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin","MAIL=/var/mail/root","LOGNAME=root","USER=root","HOME=/root","SHELL=/bin/bash","SUDO_COMMAND=/usr/bin/echo bla","SUDO_USER=root","SUDO_UID=0","SUDO_GID=0"]}}}
```

If you comment out the "x.sudo" expression you can see via valgrind that the struct json_object is leaked. 

The top-level JSON is freed, but the embedded ones are not.

**Comment by bazsi** [comment]
I've fixed the leak by using the FilterX weakref mechanism. This does not seem to leak anymore.

**Comment by kira-syslogng** [comment]
Build FAILURE

File: modules/json/json-plugin.c (hunk 1 above)
Comment (bazsi) [reply modules/json/json-plugin.c]: Just pushed a separate patch to rename this, we can revert that as a whole if you don't agree.

**Review by bazsi — APPROVED** [review]
I've removed the huge Makefile.am change, so now it only contains the fixes I requested against the branch originally and then fixed them up myself.

**Pull request merged**